modules/oe_translation_cdt/src/OeTranslationCdtServiceProvider.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Drupal\oe_translation_cdt;
+
+use Drupal\Core\DependencyInjection\ContainerBuilder;
+use Drupal\Core\DependencyInjection\ServiceProviderBase;
+use Drupal\Core\Site\Settings;
+use OpenEuropa\CdtClient\ApiClient;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Registers a CDT Client service with custom config.
+ */
+class OeTranslationCdtServiceProvider extends ServiceProviderBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function register(ContainerBuilder $container): void {
```

Comment:
I was expecting a single service class that handles instantiation and authentication, but I dont mind this approach

Reply:
Creating a wrapper class with a `->getClient()` factory was my first thought.

But then I started writing tests, and it turned out that having the client as a standalone service simplifies the use of PHPUnit prophecies. I can just create a dummy object and point it to the container. All other services use it from that moment on.

Having an additional layer of logic can complicate things a bit. For example, if I have an `ApiWrapper` service that creates an `ApiClient` and authenticates it, I would have to prophesize its `getClient()` and set the dummy `ApiWrapper` service to the container. If I do this, I won't be able to test the authentication methods of the service because they will also become dummy. According to the Prophecy docs, we should separate classes so that they are testable (classic SRP rule).

The service provider is necessary because I need to read the static settings and pass them to the constructor. I can't do this using the services.yml file.

**However**, after having some reading, I could make few improvements to this code 🙂

1. I can create a classic `ApiWrapper` class that provides the `getClient()` method that gets and authenticates the client. The `ApiClient` is passed to the service via dependency injection.
2. The `ApiClient` will be defined in `*.services.yml` as `public: false`, so it can only be a dependency. It will make use of container parameters to set up the constructor variables.

```yml
services:
  oe_translation_cdt.api_wrapper:
    class: Drupal\oe_translation_cdt\Api\CdtApiWrapper
    arguments: ['@oe_translation_cdt.api_client', '@http_client', '@state', '@datetime.time']
    
  oe_translation_cdt.api_client:
    class: OpenEuropa\CdtClient\ApiClient
    arguments:
      - @ http_client
      - @ psr17.server_request_factory
      - @ psr17.stream_factory
      - @ oe_translation_cdt.psr17.uri_factory
      - %cdt.client_configuration%
```

3. I define a new compiler pass, and set the `cdt.client_configuration` parameter to the container. This is how core does it.